### PR TITLE
update `status` and `result` fields' transformation in domain layer

### PR DESCRIPTION
**Pull request by d4x1**
### ⚠️ Pre Checklist

> Please complete _ALL_ items in this checklist, and remove before submitting

- [x] I have read through the [Contributing Documentation](https://devlake.apache.org/community/).
- [x] I have added relevant tests.
- [x] I have added relevant documentation.
- [x] I will add labels to the PR, such as `pr-type/bug-fix`, `pr-type/feature-development`, etc.

### Summary
What does this PR do?
Devlake team update `result` and `status` fields' transformation rules. 

### Does this close any open issues?
Related to #6139

### Screenshots
Include any relevant screenshots here.

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: backend/plugins/bamboo/tasks/deploy_build_convertor.go [hunk 1]
```diff
@@ -113,14 +113,15 @@ func ConvertDeployBuilds(taskCtx plugin.SubTaskContext) errors.Error {
...
 				}, input.DeploymentState),
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					Done:       []string{"Finished", "FINISHED"},
-					NotStarted: []string{"not_built", "NOT_BUILT", "Not_Built", "PENDING", "QUEUED"},
-					Default:    devops.STATUS_IN_PROGRESS,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusFinished},
+					InProgress: []string{StatusInProgress, StatusPending, StatusQueued},
+					Other:      []string{StatusNotBuilt},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/bamboo/tasks/deploy_build_convertor.go:124]: 1. Other is the same as Default

File: backend/plugins/bamboo/tasks/job_build_convertor.go [hunk 2]
```diff
@@ -72,15 +72,16 @@ func ConvertJobBuilds(taskCtx plugin.SubTaskContext) errors.Error {
...
 
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					Done:       []string{"Finished", "FINISHED"},
-					NotStarted: []string{"not_built", "NOT_BUILT", "Not_Built", "PENDING", "QUEUED"},
-					Default:    devops.STATUS_IN_PROGRESS,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusFinished},
+					InProgress: []string{StatusInProgress, StatusPending, StatusQueued},
+					Other:      []string{StatusNotBuilt},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/bamboo/tasks/job_build_convertor.go:84]: 1

File: backend/plugins/bamboo/tasks/plan_build_convertor.go [hunk 3]
```diff
@@ -70,15 +70,16 @@ func ConvertPlanBuilds(taskCtx plugin.SubTaskContext) errors.Error {
...
 
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					Done:       []string{"Finished", "FINISHED"},
-					NotStarted: []string{"not_built", "NOT_BUILT", "Not_Built", "PENDING", "QUEUED"},
-					Default:    devops.STATUS_IN_PROGRESS,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusFinished},
+					InProgress: []string{StatusInProgress, StatusPending, StatusQueued},
+					Other:      []string{StatusNotBuilt},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/bamboo/tasks/plan_build_convertor.go:82]: 1

File: backend/plugins/bitbucket/tasks/deployment_convertor.go [hunk 4]
```diff
@@ -87,13 +87,15 @@ func ConvertDeployments(taskCtx plugin.SubTaskContext) errors.Error {
...
+					Default: devops.RESULT_DEFAULT,
 				}, bitbucketDeployment.Status),
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					Done:    []string{models.COMPLETED},
-					Default: bitbucketDeployment.Status,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{models.COMPLETED, models.SUCCESSFUL, models.FAILED, models.STOPPED},
+					InProgress: []string{models.IN_PROGRESS},
+					Other:      []string{models.UNDEPLOYED},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/bitbucket/tasks/deployment_convertor.go:98]: 1

File: backend/plugins/bitbucket/tasks/pipeline_convertor.go [hunk 5]
```diff
@@ -88,16 +88,15 @@ func ConvertPipelines(taskCtx plugin.SubTaskContext) errors.Error {
...
+					Default: devops.RESULT_DEFAULT,
 				}, bitbucketPipeline.Result),
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					InProgress: []string{models.IN_PROGRESS, models.PENDING, models.BUILDING},
-					Default:    bitbucketPipeline.Status,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{models.COMPLETED, models.SUCCESSFUL, models.PASSED, models.FAILED, models.ERROR, models.STOPPED, models.HALTED},
+					InProgress: []string{models.IN_PROGRESS, models.PENDING, models.RUNNING, models.PAUSED, models.BUILDING},
+					Other:      []string{models.EXPIRED},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/bitbucket/tasks/pipeline_convertor.go:99]: 1

File: backend/plugins/gitlab/tasks/deployment_convertor.go [hunk 6]
```diff
@@ -94,18 +94,15 @@ func ConvertDeployment(taskCtx plugin.SubTaskContext) errors.Error {
...
-					Done:       []string{"COMPLETED", "UNDEPLOYED", "failed", "success", "canceled"},
-					InProgress: []string{"running"},
-					NotStarted: []string{"created"},
-					Manual:     []string{"blocked"},
-					Default:    gitlabDeployment.Status,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusSuccess, StatusCompleted, StatusFailed},
+					InProgress: []string{StatusRunning, StatusCanceled, StatusBlocked},
+					Other:      []string{StatusCreated, StatusUndeployed},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/gitlab/tasks/deployment_convertor.go:105]: 1

File: backend/plugins/gitlab/tasks/job_convertor.go [hunk 7]
```diff
@@ -85,17 +85,15 @@ func ConvertJobs(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
 				}, gitlabJob.Status),
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					InProgress: []string{"created", "waiting_for_resource", "preparing", "pending", "running", "scheduled"},
-					Manual:     []string{"manual"},
-					Default:    devops.STATUS_DONE,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusSuccess, StatusCompleted, StatusFailed},
+					InProgress: []string{StatusRunning, StatusWaitingForResource, StatusPreparing, StatusPending},
+					Other:      []string{StatusManual, StatusCreated, StatusScheduled, StatusUndeployed},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/gitlab/tasks/job_convertor.go:96]: 1

File: backend/plugins/gitlab/tasks/pipeline_convertor.go [hunk 8]
```diff
@@ -86,15 +86,15 @@ func ConvertPipelines(taskCtx plugin.SubTaskContext) errors.Error {
...
+					Default: devops.RESULT_DEFAULT,
 				}, gitlabPipeline.Status),
-				Status: devops.GetStatus(&devops.StatusRule[string]{
-					InProgress: []string{"created", "waiting_for_resource", "preparing", "pending", "running", "manual", "scheduled"},
-					Default:    devops.STATUS_DONE,
+				Status: devops.GetStatus(&devops.StatusRule{
+					Done:       []string{StatusSuccess, StatusCompleted, StatusFailed, StatusCanceled},
+					InProgress: []string{StatusRunning, StatusWaitingForResource, StatusPending, StatusPreparing},
+					Other:      []string{StatusCreated, StatusManual, StatusScheduled, StatusSkipped, StatusUndeployed},
+					Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/gitlab/tasks/pipeline_convertor.go:97]: 1

File: backend/plugins/jenkins/tasks/stage_convertor.go [hunk 9]
```diff
@@ -100,21 +100,21 @@ func ConvertStages(taskCtx plugin.SubTaskContext) (err errors.Error) {
...
+				Success: []string{SUCCESS},
+				Failed:  []string{FAILED, FAILURE, ABORTED},
+				Default: devops.RESULT_DEFAULT,
+			}, body.Status)
+
+			jenkinsTaskStatus := devops.GetStatus(&devops.StatusRule{
+				Done:       []string{SUCCESS, FAILURE, FAILED, ABORTED},
+				InProgress: []string{},
+				Other:      []string{NOT_BUILD, UNSTABLE},
+				Default:    devops.STATUS_OTHER,
```
Comment (abeizn) [inline backend/plugins/jenkins/tasks/stage_convertor.go:113]: 1

File: backend/core/models/domainlayer/devops/cicd_pipeline.go [hunk 10]
```diff
@@ -46,71 +46,80 @@ func (CICDPipeline) TableName() string {
...
-type StatusRule[T comparable] struct {
-	InProgress []T
-	NotStarted []T
-	Done       []T
-	Manual     []T
+
+type StatusRule struct {
+	InProgress []string
+	Done       []string
+	Other      []string
```
Comment (abeizn) [inline backend/core/models/domainlayer/devops/cicd_pipeline.go:68]: please remove it

File: backend/core/models/domainlayer/devops/cicd_pipeline.go [hunk 11]
```diff
@@ -46,71 +46,80 @@ func (CICDPipeline) TableName() string {
...
 	Default    string
 }
 
-func caseInSensitiveEqual(src string, dst string) bool {
-	return strings.EqualFold(src, dst)
+type StatusRuleCommon[T comparable] struct {
+	InProgress []T
+	Done       []T
+	Other      []T
+	Default    string
```
Comment (abeizn) [inline backend/core/models/domainlayer/devops/cicd_pipeline.go:76]: please remove it

File: backend/plugins/bamboo/tasks/deploy_build_convertor.go (hunk 1 above)
Comment (d4x1) [reply backend/plugins/bamboo/tasks/deploy_build_convertor.go:123]: Got it. I'll fix it later.

**Review by abeizn — APPROVED** [review]

**Pull request merged**